### ISSUE TITLE
How Can I finetune the deeplabv3+ on ms coco dataset

### ISSUE THREAD

**Issue by feixuedudiao** [tensorflow/models#4795]
Please go to Stack Overflow for help and support:

http://stackoverflow.com/questions/tagged/tensorflow

### **System information**
- **What is the top-level directory of the model you are using**: deeplab
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:ubuntu 16.04
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: tf1.8
- **Bazel version (if compiling from source)**:
- **CUDA/cuDNN version**:cuda9.0/cuDNN7.15
- **GPU model and memory**: titan xp * 4 12G * 4 
- **Exact command to reproduce**:

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

### Describe the problem
I am finetuning the deeplabv3+ using the mobilenetv2 model with pretrain  imagenets, on the ms coco dataset. I change some code in the scripts.
segmentation_dataset.py:
`_MS_COCO_SEG_INFORMATION = DatasetDescriptor(
    splits_to_sizes={
        'train': 118286,
        'val': 5000,
        #'trainval': 118286,
    },
    num_classes=81,
    ignore_label=255,
)`

_DATASETS_INFORMATION = {
    'cityscapes': _CITYSCAPES_INFORMATION,
    'pascal_voc_seg': _PASCAL_VOC_SEG_INFORMATION,
    'ms_coco': _MS_COCO_SEG_INFORMATION,
    'ade20k': _ADE20K_INFORMATION,
}

train_utils.py:
change   the code
`exclude_list = ['global_step']` 
with
`  exclude_list = ['global_step','logits']'

and set the variables 
flags.DEFINE_boolean('initialize_last_layer', False,
                     'Initialize the last layer.')

flags.DEFINE_boolean('last_layers_contain_logits_only', True,
                     'Only consider logits as last layers or not.')

but when I run the script trian.py ,it has no problem; when I run the script eval.py, it report :
assertion failed: [`predictions` out of bound] [Condition x < y did not hold element-wise:] [x (mean_iou/confusion_matrix/control_dependency_1:0) = ] [0 0 0...] [y (mean_iou/ToInt64_2:0) = ] [81]

what's wrong? who can tell me? thinks.

**Comment by yrd241** [comment]
The num_classes you set to _MS_COCO_SEG_INFORMATION is 81 which means you have a list of labels that max entry is 80, however you get a label[81] as the error invoked. I think maybe something wrong in the setting of your dataset's num_classes.

**Comment by Ekko1992** [comment]
@feixuedudiao I'm wondering how did you use MS coco dataset while finetuning deeplab models? which script did you use to convert the dataset to tfrecord? Could you please provide any suggestions, thx.

**Comment by tensorflowbutler** [comment]
Hi There,
 We are checking to see if you still need help on this, as this seems to be considerably old issue. Please update this issue with the latest information, code snippet to reproduce your issue and error you are seeing.
 If we don't hear from you in the next 7 days, this issue will be closed automatically. If you don't need help on this issue any more, please consider closing this.

**Comment by bsguru84** [comment]
> train_utils.py:
> change the code
> `exclude_list = ['global_step']`
> with
> ` exclude_list = ['global_step','logits']'
> 
> and set the variables
> flags.DEFINE_boolean('initialize_last_layer', False,
> 'Initialize the last layer.')
> 
> flags.DEFINE_boolean('last_layers_contain_logits_only', True,
> 'Only consider logits as last layers or not.')

Hi , I am also planning to do same task.
Can I ask why we need below changes:

train_utils.py:
change the code
exclude_list = ['global_step']
with
` exclude_list = ['global_step','logits']'

and set the variables
flags.DEFINE_boolean('initialize_last_layer', False,
'Initialize the last layer.')

flags.DEFINE_boolean('last_layers_contain_logits_only', True,
'Only consider logits as last layers or not.')

BTW finally did it work for you?

**Comment by YknZhu** [comment]
> > train_utils.py:
> > change the code
> > `exclude_list = ['global_step']`
> > with
> > ` exclude_list = ['global_step','logits']'
> > and set the variables
> > flags.DEFINE_boolean('initialize_last_layer', False,
> > 'Initialize the last layer.')
> > flags.DEFINE_boolean('last_layers_contain_logits_only', True,
> > 'Only consider logits as last layers or not.')
> 
> Hi , I am also planning to do same task.
> Can I ask why we need below changes:
> 
> train_utils.py:
> change the code
> exclude_list = ['global_step']
> with
> ` exclude_list = ['global_step','logits']'
> 
> and set the variables
> flags.DEFINE_boolean('initialize_last_layer', False,
> 'Initialize the last layer.')
> 
> flags.DEFINE_boolean('last_layers_contain_logits_only', True,
> 'Only consider logits as last layers or not.')
> 
> BTW finally did it work for you?

basically if we fine-tune on a dataset with different # of classes, the final layer that produces per-class logits needs to be trained from scratch. that is why we should set `initialize_last_layer=False` to not initialize logits layer from checkpoint (as the number of parameters will not match). 

If that flag is set, it should add logit layers into exclude list per https://github.com/tensorflow/models/blob/master/research/deeplab/utils/train_utils.py#L211
